### PR TITLE
Add "free mod" user mod selection support to the playlists system

### LINKED ISSUE #11737: Free mods are selectable when creating a playlist, but not when playing
There's no button to make a selection.

### PR DESCRIPTION
**Pull request by peppy**
A bit of an oversight when implementing. The available user mods for selection are not shown anywhere on playlist items yet, but I think that can come as a later step (once we have a better design for the playlists cards which can hold more information, such as the "maximum attempts" count as well).

Not 100% sure on having the selection region constructed in each of the screen implementations (the only difference between them is some padding). Could potentially reverse the direction and have it created in the base screen and inserted in the correct place for each usage if that sounds better.

Closes https://github.com/ppy/osu/issues/11737.

### CODE REVIEW

File: osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs [hunk 1]
```diff
@@ -131,6 +177,18 @@ private void selectedItemChanged()
             UpdateMods();
 
             Ruleset.Value = SelectedItem.Value.Ruleset.Value;
+
+            if (SelectedItem.Value?.AllowedMods.Any() != true)
```
Comment (bdach) [inline osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs:181]: `SelectedItem.Value` cannot be null at this stage due to the early null guard a few lines up. Therefore this won't hide the mod overlay if the item changes to `null`.

File: osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs [hunk 2]
```diff
@@ -430,10 +389,8 @@ protected override void Dispose(bool isDisposing)
                 client.RoomUpdated -= onRoomUpdated;
                 client.LoadRequested -= onLoadRequested;
             }
-        }
 
-        private class UserModSelectOverlay : LocalPlayerModSelectOverlay
-        {
+            modSettingChangeTracker?.Dispose();
```
Comment (bdach) [inline osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs:393]: This is a drive-by fix for this not having been disposed before, right? Looks to be a good idea, just double-checking.

**Review by bdach — CHANGES_REQUESTED** [review]
Rest seems okay. Playtested a little and seems to behave ok.

File: osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs:181]: i think the null check may not be required then? i noticed in testing that the room doesn't go null when no selection exists, but is a placeholder empty model instead.

File: osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs:393]: is github playing tricks? i didn't add this line, it was already there.

File: osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs:181]: The selected item goes null for a short time when room settings change. This is to prevent showing an outdated beatmap while the online lookup happens.

File: osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs (hunk 2 above)
Comment (bdach) [reply osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs:393]: Interesting, seems like something went wrong there in the web diff. This does not appear as a change anymore.

File: osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs [hunk 3]
```diff
@@ -73,14 +102,31 @@ protected override void LoadComplete()
             base.LoadComplete();
 
             SelectedItem.BindValueChanged(_ => Scheduler.AddOnce(selectedItemChanged));
-            SelectedItem.Value = Playlist.FirstOrDefault();
+
+            Playlist.BindCollectionChanged(onPlaylistChanged, true);
```
Comment (smoogipoo) [inline osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs:106]: Keep in mind this is not correct anymore and will break multiplayer. Tempted to say this shouldn't exist at all.

Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs]: Yep, should be gone.

File: osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs [hunk 4]
```diff
@@ -117,17 +158,31 @@ private void selectedItemChanged()
...
-                                     .Where(m => SelectedItem.Value.AllowedMods.Any(a => m.GetType() == a.GetType()))
+                                     .Where(m => selected.AllowedMods.Any(a => m.GetType() == a.GetType()))
                                      .ToList();
 
             UpdateMods();
 
-            Ruleset.Value = SelectedItem.Value.Ruleset.Value;
+            Ruleset.Value = selected.Ruleset.Value;
+
+            if (selected.AllowedMods.Any() != true)
```
Comment (smoogipoo) [inline osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs:175]:
```suggestion
            if (!selected.AllowedMods.Any())
```

**Review by smoogipoo — CHANGES_REQUESTED** [review]
Looks good otherwise

**Review by smoogipoo — APPROVED** [review]
lgtm
